torsvc/torcontrol.go

```diff
@@ -0,0 +1,138 @@
+package torsvc
+
+import (
+	"net"
+	"net/textproto"
+	"bufio"
+	"fmt"
+	"strings"
+)
+
+const (
+	addStr  = "ADD_ONION"
+	authStr = "AUTHENTICATE"
+	delStr  = "DEL_ONION"
+	success = 250
+)
+
+type TorControl struct {
+	conn     net.Conn
+	reader   *textproto.Reader
+	Password string
+	Port     string
+	TargPort string
+	VirtPort string
+	PrivKey  string
+	Save     bool
+}
+
+// AuthWithPass authenticates via password to Tor's ControlPort.
+//
+// Note: AuthWithPass must be called even if the ControlPort has no
+// authentication mechanism in place.
+func (tc *TorControl) AuthWithPass() error {
+	_, _, err := tc.sendCommand(authStr + " \"" + tc.Password + "\"\n")
+	return err
+}
+
+// AddOnion creates a Tor v2 hidden service. This hidden service is available as
+// long as the connection to Tor's ControlPort is kept open.
+func (tc *TorControl) AddOnion() (string, error) {
+	var command string
+
+	// If a private key is provided to the AddOnion command, we use it instead
+	// of creating a new v2 hidden service.
+	if tc.PrivKey != "" {
+		command = addStr + " RSA1024:" + tc.PrivKey
+	} else {
+		command = addStr + " NEW:RSA1024"
+	}
+
+	// If we are not going to save this private key, set the DiscardPk flag.
+	if !tc.Save {
+		command += " Flags=DiscardPk"
+	}
+
+	// Add the VIRTPORT and TARGET ports to the command.
+	command += " Port=" + tc.VirtPort + "," + tc.TargPort + "\n"
+
+	// Send the command to Tor's ControlPort.
+	_, message, err := tc.sendCommand(command)
+	if tc.Save {
+		// TODO(eugene)
+		// Parse out the private key from response and store it somewhere.
+	}
+
+	// Parse out the hidden service from the response and return it.
+	i := strings.Index(message, "ServiceID")
```

Comment:
The single letter variable names could use a bit more descriptive variable names. Also adding some inline comments showing at a high-level what the response looks like would greatly aide in the readability of this section.

Here's a section from the control port spec:
```
     C: ADD_ONION NEW:BEST Port=22 Port=80,8080
     S: 250-ServiceID=testonion1234567
     S: 250-PrivateKey=RSA1024:[Blob Redacted]
     S: 250 OK
```